### PR TITLE
refactor justificiation config(golang)

### PR DESCRIPTION
**Pull request by capri-xiyue**
fix golang part of  #280

### CODE REVIEW

File: clients/go/apis/v1alpha1/config.go [hunk 1]
```diff
@@ -358,3 +349,24 @@ func (r *AuditRule) SetDefault() {
 		r.LogType = AuditLogRequest_DATA_ACCESS.String()
 	}
 }
+
+// Justification specifies the config used to integrate with JVS.
+type Justification struct {
+	// JVSPublicKeysEndpoint is the endpoint where public keys may be retrieved from the JVS.
+	// These will be used to validate JWT tokens that are passed in through the
+	// "justification-token" header.
+	JVSPublicKeysEndpoint string `yaml:"jvs_public_keys_endpoint,omitempty"`
```
Comment (yolocs) [inline clients/go/apis/v1alpha1/config.go:358]: I think JVS here is redundant info. Should we remove it?

File: clients/go/apis/v1alpha1/config.go [hunk 2]
```diff
@@ -358,3 +349,24 @@ func (r *AuditRule) SetDefault() {
...
+	// all manual calls are expected to pass in a justification in the "justification-token" header.
+	// If omitted, justifications will not be added to logs, even if provided.
+	Enabled bool `yaml:"enabled,omitempty"`
+}
+
+// Validate validates the Justification.
+func (justification *Justification) Validate() error {
+	if justification.Enabled && justification.JVSPublicKeysEndpoint == "" {
+		return fmt.Errorf("jvs_public_keys_endpoint must be specified when justification is enabled")
+	}
```
Comment (yolocs) [inline clients/go/apis/v1alpha1/config.go:370]: nit. I recommend do the following. If we need to add configs in the future.

```suggestion
	if !justification.Enabled {
		return nil
	}
	if justification.JVSPublicKeysEndpoint == "" {
		return fmt.Errorf("jvs_public_keys_endpoint must be specified when justification is enabled")
	}
```

File: clients/go/apis/v1alpha1/config_test.go [hunk 3]
```diff
@@ -149,180 +152,226 @@ func TestValidate(t *testing.T) {
 		name    string
 		cfg     *Config
 		wantErr string
-	}{{
```
Comment (yolocs) [inline clients/go/apis/v1alpha1/config_test.go:152]: Please try to keep the current bracelets arrangement to avoid causing unnecessary changes appearing.

File: clients/go/pkg/auditopt/config.go [hunk 4]
```diff
@@ -123,8 +123,8 @@ func interceptorFromConfigFile(ctx context.Context, path string, lookuper envcon
 			return fmt.Errorf("SecurityContext must be provided to use interceptor")
 		}
 
-		if cfg.RequireJustification {
-			jvsClient, err := client.NewJVSClient(ctx, &client.JVSConfig{JVSEndpoint: cfg.JVSEndpoint})
+		if cfg.Justification != nil && cfg.Justification.Enabled {
+			jvsClient, err := client.NewJVSClient(ctx, &client.JVSConfig{JVSEndpoint: cfg.Justification.JVSPublicKeysEndpoint})
```
Comment (yolocs) [inline clients/go/pkg/auditopt/config.go:127]: We now should move this to `clientFromConfig` func since it's no longer gRPC specific?

**Review by yolocs — CHANGES_REQUESTED** [review]

File: clients/go/apis/v1alpha1/config.go (hunk 1 above)
Comment (capri-xiyue) [reply clients/go/apis/v1alpha1/config.go:358]: done

File: clients/go/apis/v1alpha1/config.go (hunk 2 above)
Comment (capri-xiyue) [reply clients/go/apis/v1alpha1/config.go:370]: I changed it to 
```
	if justification.Enabled {
		if justification.PublicKeysEndpoint == "" {
			return fmt.Errorf("public_keys_endpoint must be specified when justification is enabled")
		}
	}
	return nil
```
I feel this one looks more straightforward

File: clients/go/apis/v1alpha1/config_test.go (hunk 3 above)
Comment (capri-xiyue) [reply clients/go/apis/v1alpha1/config_test.go:152]: The format of this files was not standard, therefore I changed it. I didn't do it manaully, I ran the code formatter in IDE. Therefore, I believe no business logic has been changed.

File: clients/go/pkg/auditopt/config.go (hunk 4 above)
Comment (capri-xiyue) [reply clients/go/pkg/auditopt/config.go:127]: I will do the change once I work on the https://github.com/abcxyz/lumberjack/issues/283

File: clients/go/apis/v1alpha1/config.go (hunk 1 above)
Comment (sethvargo) [inline clients/go/apis/v1alpha1/config.go:358]: These seem to be missing env keys

File: clients/go/apis/v1alpha1/config.go [hunk 5]
```diff
@@ -65,15 +65,8 @@ type Config struct {
...
-	// These will be used to validate JWT tokens that are passed in through the
-	// "justification-token" header.
-	JVSEndpoint string `yaml:"jvs_endpoint,omitempty" env:"JVS_ENDPOINT,overwrite,default=localhost:8080"`
-
-	// RequireJustification enables adding justification information to audit logs. If this is enabled,
-	// all manual calls are expected to pass in a justification in the "justification-token" header.
-	// If omitted, justifications will not be added to logs, even if provided.
-	RequireJustification bool `yaml:"require_justification,omitempty" env:"REQUIRE_JUSTIFICATION,overwrite,default=false"`
+	// Justification specifies the config used to integrate with JVS.
+	Justification *Justification `yaml:"justification,omitempty" env:",noinit"`
```
Comment (sethvargo) [inline clients/go/apis/v1alpha1/config.go:69]: Why `noinit`? We have the `Enabled` field to control enablement.

File: clients/go/pkg/auditopt/config.go (hunk 4 above)
Comment (yolocs) [reply clients/go/pkg/auditopt/config.go:127]: Ack

File: clients/go/apis/v1alpha1/config_test.go (hunk 3 above)
Comment (yolocs) [reply clients/go/apis/v1alpha1/config_test.go:152]: I don't think standard go formatter will change this. Please adjust your IDE so future PRs won't have such unintended changes. Thanks.

File: clients/go/apis/v1alpha1/config.go (hunk 1 above)
Comment (yolocs) [reply clients/go/apis/v1alpha1/config.go:358]: +1

File: clients/go/apis/v1alpha1/config_test.go (hunk 3 above)
Comment (capri-xiyue) [reply clients/go/apis/v1alpha1/config_test.go:152]: For example, you can see the format of this file is different from the format of  file https://github.com/abcxyz/lumberjack/blob/main/clients/go/apis/v1alpha1/auditlogutil_test.go

Comment (sethvargo) [reply clients/go/apis/v1alpha1/config_test.go:152]: The linter should have caught this change when it was submitted. It's possible these changes came in _before_ the linter was connected.

Comment (yolocs) [reply clients/go/apis/v1alpha1/config_test.go:152]: @capri-xiyue That's not what I meant. Both styles of bracelet are technically acceptable by gofmt. I don't see a need to force converge to one style.

@sethvargo Are you sure linter will report this as an issue?

File: clients/go/apis/v1alpha1/config.go (hunk 1 above)
Comment (capri-xiyue) [reply clients/go/apis/v1alpha1/config.go:358]: Do we want to change the `PublicKeyEndpont` via env variables?
(To clarify, `PublicKeyEndpont`  is field of struct [justification] https://github.com/abcxyz/lumberjack/pull/284/files#diff-be7b293b39821f8b46a4da6a72de39a29d1508d34745442b19d0028a6dbe00bcR69)

Comment (yolocs) [reply clients/go/apis/v1alpha1/config.go:358]: I think my rule of thumb is to allow config setting/override with env var if possible. We only don't support env var for some repeated fields, e.g. rules.

File: clients/go/apis/v1alpha1/config.go (hunk 5 above)
Comment (capri-xiyue) [reply clients/go/apis/v1alpha1/config.go:69]: If user doesn't specify the justification, I prefer lumberjack initialize it to nil instead of empty struct. And it looks like this is the convention in the codebase, see https://github.com/abcxyz/lumberjack/blob/main/clients/go/apis/v1alpha1/config.go#L40-L50

File: clients/go/apis/v1alpha1/config.go (hunk 1 above)
Comment (capri-xiyue) [reply clients/go/apis/v1alpha1/config.go]: done

File: clients/go/apis/v1alpha1/config_test.go (hunk 3 above)
Comment (sethvargo) [reply clients/go/apis/v1alpha1/config_test.go:152]: Nope :)

File: clients/go/apis/v1alpha1/config.go (hunk 5 above)
Comment (sethvargo) [reply clients/go/apis/v1alpha1/config.go:69]: Sure, but if you let envconfig initialize the struct, `Enabled` will default to `false`. Then you don't have to check `justification != nil && justification.Enabled`, you can just always check `justification.Enabled` instead.

Comment (capri-xiyue) [reply clients/go/apis/v1alpha1/config.go:69]: I think it could be useful and safer if we can differentiate among unset(disable justification), user enable the justification, user disable the justification explicitly. In addition,  the codebase already uses `nolint` for other configs like `Backend`, `Condition` and `SecurityContext`. I prefer to keep the code style consistent.

**Review by yolocs — APPROVED** [review]

File: clients/go/apis/v1alpha1/config.go (hunk 5 above)
Comment (capri-xiyue) [reply clients/go/apis/v1alpha1/config.go:69]: I will merge this PR for now since https://github.com/abcxyz/lumberjack/issues/283 depends on it. I don't have strong preference against nil or empty struct initialization. If we think empty struct is better, I will change all configs including `Backend`, `Condition` and `SecurityContext` to make the code style consistent in another PR.
